kb/2025-05-26/velero-backup-restore.md

```diff
@@ -0,0 +1,358 @@
+```
+
+## Preparing the Virtual Machine and Image
+
+Refer to the Harvester documentation on how to perform the following steps using the NFS CSI storage class:
+
+ * [Create a virtual machine image](https://docs.harvesterhci.io/v1.6/advanced/csidriver#virtual-machine-image-creation)
+ * [Create a virtual machine](https://docs.harvesterhci.io/v1.6/advanced/csidriver#virtual-machine-creation)
+
+This example uses the [Ubuntu 24.04 raw image](https://cloud-images.ubuntu.com/minimal/releases/noble/) to create a virtual machine with NFS root and data volumes in the custom `demo-src` namespace:
+
+![image](vm-create.png)
+
+:::note
+
+For CSI snapshotting to work, the NFS volumes must have the `volumeMode` set to `Filesystem`.
+
+:::
+
+Access the virtual machine via SSH, and add some files to both the root and data volumes.
+
+The data volume needs to be partitioned, with a file system created and mounted before files can be written to it.
+
+## Backup the Source Namespace
+
+Use the `velero` CLI to create a backup of the `demo-src` namespace:
+
+```sh
+BACKUP_NAME=backup-demo-src-`date "+%s"`
+
+velero backup create "${BACKUP_NAME}" --include-namespaces demo-src
+```
+
+This creates a backup containing namespace-scoped resources like the virtual machine, its volumes, image and other associated configuration.
+
+:::important
+
+Velero only uploads Kubernetes objects configuration to the remote backup location, not the data in the CSI snapshots. For more information, refer to the [Velero CSI documentation](https://velero.io/docs/v1.15/csi/#how-it-works---overview).
+
+See also section on [Restore To A Different Cluster](#restore-to-a-different-cluster) below.
+
+:::
+
+Confirm that the backup completed successfully:
+
+```sh
+velero backup get "${BACKUP_NAME}"
+```
+
+```sh
+NAME                         STATUS      ERRORS   WARNINGS   CREATED                         EXPIRES   STORAGE LOCATION   SELECTOR
+backup-demo-src-1747954979   Completed   0        0          2025-05-22 16:04:46 -0700 PDT   29d       default            <none>
+```
+
+The `velero backup describe` and `velero backup logs` commands can be used to confirm the resources included in the backup.
+
+## Restore To A Different Namespace
+
+This section describes how to restore the backup from the `demo-src` namespace to a new namespace named `demo-dst`:
+
+```sh
+velero restore create \
+  --from-backup "${BACKUP_NAME}" \
+  --namespace-mappings "demo-src:demo-dst" \
+  --exclude-resources "virtualmachineimages.harvesterhci.io" \
+  --labels "velero.kubevirt.io/clear-mac-address=true,velero.kubevirt.io/generate-new-firmware-uuid=true"
+```
+
+ * During the restore:
+
+   * The virtual machine MAC address and firmware UUID are reset to avoid potential conflicts with existing virtual machines.
+   * The virtual machine image manifest is excluded as it is not needed in the new namespace. The restored virtual machine uses the original image.
```

Comment:
I think I understand this line about the virtual machine image, but only because we hit a upgrade where the BackingImage objects went sideways. We ended up being ok, but it was almost a bad week. 

You might want to consider explaining  (or linking to) how `vmimage`, `BackingImages`, and `BackupBackingImages` are all related or may need to be manipulated for the purpose of this process.

Reply:
Yeah, vm image requires some additional testing and consideration. Ultimately, we only need the `VirtualMachineImage`, `DataVolume` and `PersistentVolumeClaim` (of the golden image) to be included in the backup. During restore though, we need a way to coordinate the dependencies between these resources (via labeling and restore ordering) to avoid contending with CDI.

Reply:
In this example of performing restoring on the same cluster, which isn't a DR scenario, the vm image can be shared across namespaces.

Reply:
To confirm though, in this example, we are using CSI NFS, not LH, so there are no `BackingImages`, `BackupBackingImages` etc. In light of that, the only thing needed by the user is to create the `VirtualMachineImage` resource and CDI will do the rest (i.e. download from the original remote source). For upload, user will have to re-upload the image from their "local" to their target cluster, until we close the gap of backup/restore vm image.